rfc/0000-keyless-job-auth.rst

```diff
@@ -0,0 +1,204 @@
+   |<-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+   |                                                                                        |                                                      |                                      |
+
+
+It is worth emphasizing that the purpose of this feature is *not* to provide a
+fully complete and compliant metadata server to Hail Batch. Rather, the aim is to provide the
+minimum functionality necessary to allow Hail libraries and popular first-party
+tools like `gcloud` and `az` the ability to obtain short-lived credentials without
+exposing key files to user code. As such, an implementation may implement just the
+endpoints necessary to run the below examples for at least one version of `gcloud`/`az`
+and all supported versions of `hail`.
+
+
+Examples
+--------
+
+Under the proposed change, the following Batch job commands should succeed:
+
+.. code-block:: python
+   j.command('gcloud storage ls <MY_BUCKET>')
+   j.command('hailctl batch submit <MY_SCRIPT>')
+
+
+Effect and Interactions
+-----------------------
+
+This change adds a method through which jobs can obtained short-lived access
+tokens without directly accessing long-lived credentials. With such a solution
+in place, we can eventually remove the long-lived credentials from job containers,
+mitigating the risk of exfiltration.
+
+So long as the firewall rules are correctly configured to redirect metadata traffic
+to the Batch worker, there should be no adverse interactions with the existing
+system as such traffic was previously forbidden.
+
+
+Inspiration & Alternatives
+--------------------------
+
+We can look to the Kubernetes project for examples of integrating with cloud
+identity providers. In particular, we will examine [GKE's Workload Identity](https://cloud.google.com/kubernetes-engine/docs/concepts/workload-identity#what_is).
+Workload identity allows pods to obtain credentials for GCP IAM identities. To
+do so, GKE runs the [GKE Metadata Server](https://cloud.google.com/kubernetes-engine/docs/concepts/workload-identity#metadata_server)
+which functions similarly to what is described in this proposal.
+
+The difference arises in how the metadata server fulfills the user's request for
+an access token. Unlike in this proposal, GKE nodes do not hold IAM credentials.
+Instead, it uses OIDC to "trade" a Kubernetes Service Account credential for a
+[preconfigured IAM credential](https://cloud.google.com/kubernetes-engine/docs/concepts/workload-identity#credential-flow). This has the advantage of not needing to distribute IAM
+credentials in GKE and enabling fine-grained mapping between GKE and IAM identities.
+However, OIDC is not easily applicable in Hail Batch because Batch is not an
+identity provider. We could remove the storage and distribution of key files in GCP by
```

Comment:
The text as written is fine but I'll note that we *could*, in theory, be an identity provider for the robot identities. That would, in theory, allow seamless cross-cloud access to resources. In practice, `gcloud` does not appear to support using OIDC credentials. The OIDC situation in general seems very new.